OSSM-3661: [DOC] Document configuring OSSM with OpenShift (or external) Prometheus (#61854)
openshift/openshift-docs: modules/ossm-integrating-with-user-workload-monitoring.adoc

////
Module included in the following assemblies:
* service_mesh/v2x/ossm-observability.adoc
////

:_content-type: PROCEDURE
[id="ossm-integrating-with-user-workload-monitoring_{context}"]
= Integrating with user-workload monitoring

> By default, {SMProductName} (OSSM) installs the Service Mesh Control Plane (SMCP) with a dedicated instance of Prometheus for collecting metrics from a mesh. However, production systems need more advanced monitoring systems, like {product-title} monitoring for user-defined projects.

Reviewer: I wonder if "control plane" should be lowercase here, since it's a component as opposed to a product?
It's lowercase in the rest of the doc.

The following steps show how to integrate Service Mesh with user-workload monitoring.

.Prerequisites

* User-workload monitoring is enabled.

> * OSSM Operator 2.4 is installed.

Reviewer: I'd use the actual full name of the Operator here to avoid confusion

.Procedure

. Configure the SMCP for external Prometheus:
+
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  addons:
    prometheus:
      enabled: false # <1>
    grafana:
      enabled: false # <2>
    kiali:
      enabled: false
----
<1> Disable the default Prometheus instance provided by OSSM.
<2> Disable Grafana. It is not supported with an external Prometheus instance.

> . Apply a custom network policy to allow ingress traffic from the `user-workload monitoring` namespace:

Reviewer: Is `user-workload monitoring` the actual name of the namespace, or should this be "`user-workload' monitoring namespace"? There's no literal reference to the namespace in the YAML, so it's hard to tell, but IIRC I haven't ever seen a namespace with a space in the name, it's usually dashes I think

Author: Looking at https://docs.openshift.com/container-platform/4.12/monitoring/enabling-monitoring-for-user-defined-projects.html I now wonder if it's `openshift-user-workload-monitoring` so I will chechk.

Reviewer 2: The monitoring namespace is selected via a label selector (`network.openshift.io/policy-group: monitoring`), not via a direct reference, so it may be more accurate to just call it the "monitoring namespace". Currently, I believe at least the namespaces `openshift-monitoring` and `openshift-user-workload-monitoring` have this label. I _want_ to say the only namespace that _needs_ to access the workload namespace (`bookinfo` in this case) would be `openshift-user-workload-monitoring`, in which case it would be accurate to just say `openshift-user-workload-monitoring`, but I'm not 100% sure on that.

Either way, I think just referring to it as the "monitoring namespace" here would be better.
```suggestion
. Apply a custom network policy to allow ingress traffic from the monitoring namespace:
```

:_content-type: PROCEDURE
[id="ossm-integrating-with-user-workload-monitoring_{context}"]
= Integrating with user-workload monitoring

By default, {SMProductName} (OSSM) installs the Service Mesh Control Plane (SMCP) with a dedicated instance of Prometheus for collecting metrics from a mesh. However, production systems need more advanced monitoring systems, like {product-title} monitoring for user-defined projects.

The following steps show how to integrate Service Mesh with User-Workload Monitoring.

.Procedure

. Enable monitoring in user-defined projects by adding the following configuration to the `cluster-monitoring-config` config map:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
----

. Configure the Service Mesh control plane for User-Workload Monitoring:
+
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  addons:
    prometheus:
      enabled:  false # <1>
  meshConfig:
    extensionProviders:
    - name: prometheus # <2>
      prometheus: {}
  security:

>     manageNetworkPolicy: false # <3>

Reviewer 3: > I believe our "default" for the documentation should be the custom NetworkPolicy method

Yes, I agree.

> The tl;dr reason for this is that turning off manageNetworkPolicy has broad implications

I wouldn't say so... Service Mesh users should favor Istio AuthorizationPolicy over kubernetes NetworkPolicy. But anyway, I agree that custom network policy is better than turning off mangeNetworkPolicy to avoid user concerns.

      operator: DoesNotExist
  podMetricsEndpoints:
  - path: /stats/prometheus
    interval: 30s
    relabelings:
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_container_name]
      regex: "istio-proxy"
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_annotationpresent_prometheus_io_scrape]
    - action: replace
      regex: (\d+);(([A-Fa-f0-9]{1,4}::?){1,7}[A-Fa-f0-9]{1,4})
      replacement: '[$2]:$1'
      sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_port, 
      __meta_kubernetes_pod_ip]
      targetLabel: __address__
    - action: replace
      regex: (\d+);((([0-9]+?)(\.|$)){4})
      replacement: $2:$1
      sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_port, 
      __meta_kubernetes_pod_ip]
      targetLabel: __address__
    - action: labeldrop
      regex: "__meta_kubernetes_pod_label_(.+)"
    - sourceLabels: [__meta_kubernetes_namespace]
      action: replace
      targetLabel: namespace
    - sourceLabels: [__meta_kubernetes_pod_name]
      action: replace
      targetLabel: pod_name
    - action: replace
      replacement: "<SMCP_NAME>-<SMCP_NAMESPACE>" # <2>
      targetLabel: mesh_id
----

> <1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` in all mesh namespaces, including the control plane namespace.

Reviewer 2: ```suggestion
<1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the resource to the namespace it should apply to. You can apply it to the control plane namespace to apply it for all mesh namespaces.
```
Just elaborating on this to explain that you can restrict ServiceMonitors/PodMonitors to specific namespaces if desired.

Reviewer 3: > You can apply it to the control plane namespace to apply it for all mesh namespaces.

This statement is not correct. OpenShift monitoring stack does not know anything about mesh namespaces. On the other hand, Istio does not handle Service/Pod monitors, so this semantic does not apply to monitoring resources.

= Integrating with user-workload monitoring

By default, {SMProductName} (OSSM) installs the Service Mesh Control Plane (SMCP) with a dedicated instance of Prometheus for collecting metrics from a mesh. However, production systems need more advanced monitoring systems, like {product-title} monitoring for user-defined projects.

The following steps show how to integrate Service Mesh with User-Workload Monitoring.

.Procedure

. Enable monitoring in user-defined projects by adding the following configuration to the `cluster-monitoring-config` config map:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
----

. Configure the Service Mesh control plane for User-Workload Monitoring:
+
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  addons:
    prometheus:
      enabled: false # <1>
    grafana:
      enabled: false # <2>
  meshConfig:
    extensionProviders:
    - name: prometheus # <3>
----
<1> Disable the default Prometheus instance provided by OSSM.

> <2> Disable Grafana.

Reviewer 3: What do you think about clarification why it's needed?
```suggestion
<2> Disable Grafana, because it's not supported with external Prometheus.
```

+
[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: user-workload-access
  namespace: bookinfo
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          network.openshift.io/policy-group: monitoring
  podSelector: {}
  policyTypes:
  - Ingress
----
+

. Apply a telemetry object to enable traffic metrics in Istio proxies:
+
[source,yaml]
----
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: enable-prometheus-metrics
  namespace: istio-system # <1>
spec:
  selector: # <2>
    matchLabels:
      app: bookinfo 
  metrics:
  - providers:
    - name: prometheus
----
<1> A telemetry resource created in the control plane namespace applies to all workloads in a mesh. To apply telemetry to only one namespace, create the resource in the target namespace.
<2> (Optional) Setting `spec.selector.matchLabels` applies the telemetry resource to specific workloads in the target namespace.
+

. Apply a `ServiceMonitor` to monitor the Istio control plane:
+
[source,yaml]
----
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: istiod-monitor

>   namespace: istio-system

Reviewer 3: ```suggestion
  namespace: istio-system # <1>
```

Author: @jewertow Do you want "A telemetry resource created in the control plane namespace applies to all workloads in a mesh. To apply telemetry to only one namespace, create the resource in the target namespace." to be repeated here?

Reviewer 3: No, because it's not Telemetry. This `<1>` should point to "<1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` in all mesh namespaces, including the control plane namespace." - same as PodMonitor.

////
Module included in the following assemblies:
* service_mesh/v2x/ossm-observability.adoc
////

:_content-type: PROCEDURE
[id="ossm-integrating-with-user-workload-monitoring_{context}"]
= Integrating with user-workload monitoring

By default, {SMProductName} (OSSM) installs the Service Mesh Control Plane (SMCP) with a dedicated instance of Prometheus for collecting metrics from a mesh. However, production systems need more advanced monitoring systems, like {product-title} monitoring for user-defined projects.

The following steps show how to integrate Service Mesh with user-workload monitoring.

.Prerequisites

> * You have cluster administrator permissions.

Reviewer 3: If we list prerequisites, we should mention about installation of OSSM Operator 2.4.

> * You have cluster administrator permissions.

Reviewer 3: I would remove this prerequisite, because this is only needed to have an access to `openshift-monitoring` namespace, so I would prefer to list the following requirements:
1) UWM monitoring is enabled.
2) OSSM Operator 2.4 is installed.

```suggestion
* You have cluster administrator permissions.
```

.Procedure

. Enable monitoring in user-defined projects by adding the following configuration to the `cluster-monitoring-config` config map:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
----

> . Configure the SMCP for user-workload monitoring:

Reviewer 3: Actually, this SMCP config has nothing specific for user-workload monitoring. Any Prometheus instance could monitor this SMCP instance.
```suggestion
. Configure the SMCP for external Prometheus:
```

+
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  addons:
    prometheus:
      enabled: false # <1>
    grafana:

>       enabled: false # <2>

Reviewer 3: We didn't mention about Kiali Operator in the "Prerequisites" section, so we have to disable Kiali installation as well, because Kiali is enabled by default, so if there is no Kiali Operator, reconciliation of SMCP will fail.

We don't have to add `<3>` and explain why we disable Kiali, because we will describe integration with Kiali in a subsequent PR.

    - namespaceSelector:
        matchLabels:
          network.openshift.io/policy-group: monitoring
  podSelector: {}
  policyTypes:
  - Ingress
----
+

. Apply a telemetry object to enable traffic metrics in Istio proxies:
+
[source,yaml]
----
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: enable-prometheus-metrics
  namespace: istio-system # <1>
spec:
  selector: # <2>
    matchLabels:
      app: bookinfo 
  metrics:
  - providers:
    - name: prometheus
----
<1> A telemetry resource created in the control plane namespace applies to all workloads in a mesh. To apply telemetry to only one namespace, create the resource in the target namespace.
<2> (Optional) Setting `spec.selector.matchLabels` applies the telemetry resource to specific workloads in the target namespace.
+

. Apply a `ServiceMonitor` to monitor the Istio control plane:
+
[source,yaml]
----
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: istiod-monitor
  namespace: istio-system # <1>
spec:
  targetLabels:
  - app
  selector:
    matchLabels:
      istio: pilot
  endpoints:
  - port: http-monitoring
    interval: 30s
    relabelings:
    - action: replace

>       replacement: "<SMCP_NAME>-<SMCP_NAMESPACE>"

Reviewer 3: ```suggestion
      replacement: "<SMCP_NAME>-<SMCP_NAMESPACE>" # <2>
```

////
Module included in the following assemblies:
* service_mesh/v2x/ossm-observability.adoc
////

:_content-type: PROCEDURE
[id="ossm-integrating-with-user-workload-monitoring_{context}"]
= Integrating with user-workload monitoring

By default, {SMProductName} (OSSM) installs the Service Mesh Control Plane (SMCP) with a dedicated instance of Prometheus for collecting metrics from a mesh. However, production systems need more advanced monitoring systems, like {product-title} monitoring for user-defined projects.

The following steps show how to integrate Service Mesh with User-Workload Monitoring.

.Procedure

. Enable monitoring in user-defined projects by adding the following configuration to the `cluster-monitoring-config` config map:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true

> ----

Author: I will check on this.

    extensionProviders:
    - name: prometheus # <3>
----
<1> Disable the default Prometheus instance provided by OSSM.
<2> Disable Grafana. It is not supported with an external Prometheus instance.
<3> Configure the Prometheus extension provider.
+

. Apply a custom network policy to allow ingress traffic from the `user-workload monitoring` namespace:
+
[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: user-workload-access
  namespace: bookinfo
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          network.openshift.io/policy-group: monitoring
  podSelector: {}
  policyTypes:
  - Ingress
----
+

. Apply a telemetry object to enable traffic metrics in Istio proxies:
+
[source,yaml]
----
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: enable-prometheus-metrics
  namespace: istio-system # <1>
spec:
  selector: # <2>
    matchLabels:
      app: bookinfo 
  metrics:
  - providers:
    - name: prometheus
----
<1> A telemetry resource created in the control plane namespace applies to all workloads in a mesh. To apply telemetry to only one namespace, create the resource in the target namespace.

> <2> (Optional) Setting `spec.selector.matchLabels` applies the telemetry resource to specific workloads in the target namespace.

Reviewer: ```suggestion
<2> Optional: Setting `spec.selector.matchLabels` applies the `Telemetry` object to specific workloads in the target namespace.
```
Again I'd say choose either object or resource and use consistently.

+

> . Apply a `ServiceMonitor` to monitor the Istio control plane:

Reviewer: ```suggestion
. Apply a `ServiceMonitor` object to monitor the Istio control plane:
```
Personally I think resource sounds better than object, but as mentioned above, I'd pick one and make sure it's consistent throughout the doc

+
[source,yaml]
----
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: istiod-monitor
  namespace: istio-system # <1>
spec:
  targetLabels:
  - app
  selector:
    matchLabels:
      istio: pilot
  endpoints:
  - port: http-monitoring
    interval: 30s
    relabelings:
    - action: replace

>       replacement: "<SMCP_NAME>-<SMCP_NAMESPACE>" # <2>

Reviewer: ```suggestion
      replacement: "<smcp_name>-<smcp_namespace>" # <2>
```
Format replaceable values per https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

>       targetLabel: mesh_id

Reviewer: ```suggestion
      targetLabel: <mesh_id>
```
not sure if this is a replaceable value or a literal value?
Format replaceable values per https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

----

> <1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` in all mesh namespaces, including the control plane namespace.

Reviewer: ```suggestion
<1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
```

> <2> A shared Prometheus instance can mean there are metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the `query_scope` in Kiali to only the relevant mesh metrics.

Reviewer: ```suggestion
<2> A shared Prometheus instance can include metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the query scope in Kiali to only the relevant mesh metrics.
```
Unless there's a need to include `query_scope` here I feel like it's easier to read when written as "query scope", because it makes the meaning more clear?
Just some suggestions for readability

+

> . Apply a `PodMonitor` to collect metrics from Istio proxies:

Reviewer: ```suggestion
. Apply a `PodMonitor` object to collect metrics from Istio proxies:
```

    - key: istio-prometheus-ignore
      operator: DoesNotExist
  podMetricsEndpoints:
  - path: /stats/prometheus
    interval: 30s
    relabelings:
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_container_name]
      regex: "istio-proxy"
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_annotationpresent_prometheus_io_scrape]
    - action: replace
      regex: (\d+);(([A-Fa-f0-9]{1,4}::?){1,7}[A-Fa-f0-9]{1,4})
      replacement: '[$2]:$1'
      sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_port, 
      __meta_kubernetes_pod_ip]
      targetLabel: __address__
    - action: replace
      regex: (\d+);((([0-9]+?)(\.|$)){4})
      replacement: $2:$1
      sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_port, 
      __meta_kubernetes_pod_ip]
      targetLabel: __address__
    - action: labeldrop
      regex: "__meta_kubernetes_pod_label_(.+)"
    - sourceLabels: [__meta_kubernetes_namespace]
      action: replace
      targetLabel: namespace
    - sourceLabels: [__meta_kubernetes_pod_name]
      action: replace
      targetLabel: pod_name
    - action: replace
      replacement: "<SMCP_NAME>-<SMCP_NAMESPACE>" # <2>

>       targetLabel: mesh_id

Reviewer: ```suggestion
      targetLabel: <mesh_id>
```

----

> <1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` in all mesh namespaces, including the control plane namespace.

Reviewer: ```suggestion
<1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
```

> <2> A shared Prometheus instance can mean there are metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the `query_scope` in Kiali to only the relevant mesh metrics.

Reviewer: ```suggestion
<2> A shared Prometheus instance can include metrics from multiple meshes. However, if you are deploying Kiali, the metrics must be associated with a single mesh. To associate metrics with a single mesh, add a unique `mesh_id` to each one. Doing so narrows the query scope in Kiali to only the relevant mesh metrics.
```

+

> . Go to your OpenShift Container Platform (OCP) console, and check that the metrics appear.

Reviewer: ```suggestion
. Open the {product-title} web console, and check that metrics are visible.
```

////
Module included in the following assemblies:
* service_mesh/v2x/ossm-observability.adoc
////

:_content-type: PROCEDURE
[id="ossm-integrating-with-user-workload-monitoring_{context}"]
= Integrating with user-workload monitoring

By default, {SMProductName} (OSSM) installs the Service Mesh Control Plane (SMCP) with a dedicated instance of Prometheus for collecting metrics from a mesh. However, production systems need more advanced monitoring systems, like {product-title} monitoring for user-defined projects.

The following steps show how to integrate Service Mesh with user-workload monitoring.

.Prerequisites

> * You have access to the cluster as a user with the `cluster-admin` cluster role.
> 
> * You have created the `cluster-monitoring-config ConfigMap` object.

Reviewer: ```suggestion
* You have cluster administrator permissions.
* You have created the `cluster-monitoring-config` config map.
```
A suggestion re permissions wording so that the word cluster isn't used 3 times 🙂 
+ formatting fix for the config map

> * You have created the `cluster-monitoring-config ConfigMap` object.

Reviewer 3: This point does not make sense... We should remove it or say "You have enabled user-workload monitoring", but then we should remove the first step in the "Procedure" section.

  podSelector: {}
  policyTypes:
  - Ingress
----

. Apply a `Telemetry` object to enable traffic metrics in Istio proxies:
+
[source,yaml]
----
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: enable-prometheus-metrics
  namespace: istio-system # <1>
spec:
  selector: # <2>
    matchLabels:
      app: bookinfo 
  metrics:
  - providers:
    - name: prometheus
----
<1> A `Telemetry` object created in the control plane namespace applies to all workloads in a mesh. To apply telemetry to only one namespace, create the object in the target namespace.
<2> Optional: Setting `spec.selector.matchLabels` applies the `Telemetry` object to specific workloads in the target namespace.

. Apply a `ServiceMonitor` object to monitor the Istio control plane:
+
[source,yaml]
----
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: istiod-monitor
  namespace: istio-system # <1>
spec:
  targetLabels:
  - app
  selector:
    matchLabels:
      istio: pilot
  endpoints:
  - port: http-monitoring
    interval: 30s
    relabelings:
    - action: replace
      replacement: "<smcp_name>-<smcp_namespace>" # <2>
      targetLabel: <mesh_id>
----

> <1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.

Reviewer: ```suggestion
<1> Since {product-title} monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.
```
make consistent with the earlier formatting here https://github.com/openshift/openshift-docs/pull/61854/files#diff-7e86933621f91bd09740edfa42e222d127081e41384db64c0beecd5d87ce685eR10

      operator: DoesNotExist
  podMetricsEndpoints:
  - path: /stats/prometheus
    interval: 30s
    relabelings:
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_container_name]
      regex: "istio-proxy"
    - action: keep
      sourceLabels: [__meta_kubernetes_pod_annotationpresent_prometheus_io_scrape]
    - action: replace
      regex: (\d+);(([A-Fa-f0-9]{1,4}::?){1,7}[A-Fa-f0-9]{1,4})
      replacement: '[$2]:$1'
      sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_port, 
      __meta_kubernetes_pod_ip]
      targetLabel: __address__
    - action: replace
      regex: (\d+);((([0-9]+?)(\.|$)){4})
      replacement: $2:$1
      sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_port, 
      __meta_kubernetes_pod_ip]
      targetLabel: __address__
    - action: labeldrop
      regex: "__meta_kubernetes_pod_label_(.+)"
    - sourceLabels: [__meta_kubernetes_namespace]
      action: replace
      targetLabel: namespace
    - sourceLabels: [__meta_kubernetes_pod_name]
      action: replace
      targetLabel: pod_name
    - action: replace
      replacement: "<smcp_name>-<smcp_namespace>" # <2>
      targetLabel: <mesh_id>
----

> <1> Since OpenShift Monitoring ignores the field `spec.namespaceSelector` in `ServiceMonitor` and `PodMonitor` objects, you need to apply the `PodMonitor` object in all mesh namespaces, including the control plane namespace.

Reviewer: same comment as previous callout <1>

////
Module included in the following assemblies:
* service_mesh/v2x/ossm-observability.adoc
////

:_content-type: PROCEDURE
[id="ossm-integrating-with-user-workload-monitoring_{context}"]
= Integrating with user-workload monitoring

> By default, OpenShift Service Mesh (OSSM) installs the Service Mesh Control Plane (SMCP) with a dedicated instance of Prometheus for collecting metrics from a mesh. However, production systems need more advanced monitoring systems, like OpenShift Monitoring for user-defined projects.

Reviewer 3: Could you make "OpenShift Monitoring for user-defined projects" a link to this page https://docs.openshift.com/container-platform/4.13/monitoring/enabling-monitoring-for-user-defined-projects.html?

Author: Linking directly in the body isn't possible as it breaks other systems that use this content, so I added an "Additional Resources" section and linked to "Enabling Monitoring for User-Defined Projects" from there.

The following steps show how to integrate Service Mesh with User-Workload Monitoring.

.Procedure

. Enable monitoring in user-defined projects by adding the following configuration to the `cluster-monitoring-config` config map:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    enableUserWorkload: true
----

> . Complete one of the following options to configure the Service Mesh control plane for User-Workload Monitoring.

Reviewer 3: "Complete one of the following options" - this is incorrect.
1) Disable Prometheus addon - required.
2) Configure extension provider - required.
3) Disable network policy **OR** apply custom network policy to allow ingress traffic from user-workload monitoring namespace.

** Either disable the installation of the default Prometheus instance:
+
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  addons:
    prometheus:
      enabled:  false # <1>
  meshConfig:
    extensionProviders:
    - name: prometheus # <2>
      prometheus: {}
  security:
    manageNetworkPolicy: false # <3>
----
<1> Disable the default Prometheus instance provided by OSSM.
<2> Configure the Prometheus extension provider.
<3> Disable OSSM managed network policies. 
+
[NOTE]
====
This option disables all of your OSSM managed network policies.
====
+
** Or add an additional network policy to enable ingress from `user-workload-monitoring`:
+
[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: user-workload-access
  namespace: bookinfo
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          network.openshift.io/policy-group: monitoring
  podSelector: {}
  policyTypes:

>   - Ingress

Reviewer 3: @fjglira could you modify our e2e test of integration with OpenShift Monitoring stack to use this network policy instead of disabling network policies in SMCP?

----
+
[NOTE]
====
This option lets you keep your OSSM managed network policies.
====
+

. Apply a telemetry object to enable traffic metrics in Istio proxies.
+
[source,yaml]
----
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: enable-prometheus-metrics
  namespace: istio-system
spec:
  metrics:
  - providers:
    - name: prometheus
----
+
[NOTE]
====
When a telemetry object is created in the control plane namespace, it applies to all workloads in a mesh. You can apply the telemetry object in selected namespaces or workloads by setting `spec.selector.matchLabels`. 
====

. Apply a ServiceMonitor to monitor the Istio control plane.
+
[source,yaml]
----
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: istiod-monitor
  namespace: istio-system
spec:
  targetLabels:
  - app
  selector:
    matchLabels:
      istio: pilot
  endpoints:
  - port: http-monitoring

>     interval: 30s

Reviewer 3: ```suggestion
    interval: 30s
    relabelings:
    - action: replace
      replacement: "<SMCP_NAME>-<SMCP_NAMESPACE>" <2>
      targetLabel: mesh_id
```
@fjglira could you also add this change to our e2e test and then verify all in Kiali in OCP dashboard?

Reviewer 3: @gwynnemonahan please apply this suggestion.
